app/code/ZP/LoyaltyProgram/Controller/Adminhtml/Program/MassStatus.php

```diff
@@ -3,63 +3,180 @@
 
     public function __construct(
         Context $context,
-        private Filter $filter,
-        private CollectionFactory $collectionFactory,
-        private LoyaltyProgramResource $programResource
+        LoggerInterface $logger,
+        LoyaltyProgramRepositoryInterface $programRepository,
+        CustomerProgramManagement $customerProgramManagement,
+        Helper $helper,
+        ValidatorInterface $dataValidator,
+        RequestHelper $requestHelper,
+        Filter $filter,
+        CollectionFactory $collectionFactory,
+        DataPreparer $prepareData,
     ) {
-        parent::__construct($context);
+        parent::__construct(
+            $context,
+            $logger,
+            $programRepository,
+            $customerProgramManagement,
+            $helper,
+            $dataValidator,
+            $requestHelper,
+            $filter,
+            $collectionFactory,
+            $prepareData
+        );
     }
 
     public function execute()
     {
-        $collection = $this->filter->getCollection($this->collectionFactory->create());
-        $programStatusesChanged = 0;
-        $status = (int) $this->getRequest()->getParam('status');
+        try {
+            /** @var \Magento\Backend\Model\View\Result\Redirect $resultRedirect */
+            $resultRedirect = $this->resultFactory->create(ResultFactory::TYPE_REDIRECT);
+            $this->checkProgramIds(self::EDIT);
 
-        /** @var \Magento\Backend\Model\View\Result\Redirect $resultRedirect */
-        $resultRedirect = $this->resultFactory->create(ResultFactory::TYPE_REDIRECT);
+            /** @var Collection $collection */
+            $collection = $this->filter->getCollection($this->collectionFactory->create());
+            if (!$collection->getSize()) {
+                $this->messageManager->addNoticeMessage('Selected program(s) has(have) been probably deleted!');
+            } else {
+                $this->setActiveStatusFromRequest();
 
-        try {
-            /** @var LoyaltyProgram $program */
-            foreach ($collection as $program) {
-                if (!$this->isSameStatus((bool)$status, $program)) {
-                    $program->setIsActive((bool)$status);
-                    $this->programResource->save($program);
-                    $programStatusesChanged++;
+                $this->validateCollectionPrograms($collection);
+
+                if ($this->updatedProgramsCount) {
+                    if ($this->programsForAction) {
+                        $this->programsData = $this->programsForAction;
+                        $this->beforeSave();
+                    }
+
+                    /** @var LoyaltyProgram $program */
+                    foreach ($collection->getItems() as $program) {
+                        $this->programRepository->save($program);
+                    }
+                }
+
+                if ($this->programsForAction) {
+                    $this->afterSave();
```

Comment:
No we cant, because my before and after actions which are in controllers have action logic that is not applicable to our resource model. But I will try to remember this moment and we will discuss it on our next live meeting, and may in our discussion we will find the way how to do it even more functional.